src/integer_object.cpp

```diff
@@ -223,14 +230,84 @@ Value IntegerObject::mod(Env *env, Value arg) const {
     return Value::integer(result);
 }
 
-Value IntegerObject::pow(Env *env, Value arg) const {
-    if (arg.is_fast_integer())
-        return Value::integer(::pow(m_integer, arg.get_fast_integer()));
+Value fast_pow(Env *env, nat_int_t a, nat_int_t b) {
+    if (b == 0)
+        return Value::integer(1);
+    if (b == 1)
+        return Value::integer(a);
+    if (a == 0)
+        return Value::integer(0);
 
-    arg.unguard();
-    arg->assert_type(env, Object::Type::Integer, "Integer");
-    auto result = ::pow(m_integer, arg->as_integer()->to_nat_int_t());
-    return Value::integer(result);
+    auto handle_overflow = [env, &a, &b](nat_int_t last_result, bool should_be_negative) -> Value {
+        auto result = BignumObject(a).pow(env, Value::integer(b));
+
+        // Check if the bignum pow overflowed.
+        if (result->is_float() && result->as_float()->is_infinite(env)) {
+            return result;
+        }
+
+        result = result->as_integer()->mul(env, Value::integer(should_be_negative ? -last_result : last_result));
+        return result;
+    };
+
+    nat_int_t result = 1;
+
+    bool negative = a < 0;
+    if (negative) a = -a;
+    negative = negative && (b % 2);
+
+    while (b != 0) {
+        while (b % 2 == 0) {
+            if (will_multiplication_overflow(a, a)) {
```

Comment:
can you just check instead if a > NAT_MAX_FIXNUM_SQRT?

Reply:
Yeah I guess we could do that... It would probably be more performant, right? 

My intentation was to make all overflow checks use the same method which is why I integrated this check to will_multiplication_overflow instead.

Reply:
yeah, that'd be way faster, but I still like that you separated that overflow check like that :+1:

Reply:
I changed both things you commented on. Thanks for the suggested changes 🤗